Task 03

zuwarka opened the pull request.

zuwarka committed:
Create README.md

zuwarka committed:
Create task03.rb

В данном таске выполнено три задания: привели по пять функций для структур, а также выполнили задание 3а и 3б.

zuwarka committed:
Update README.md

zuwarka committed:
Create README.md

zuwarka committed:
Create task04.rb

Задания 1 и 2. В Задании 2 есть момент, который непонятен для решения - оставлен развернутый комментарий

zuwarka committed:
Update README.md

Alex6Alex commented on homework04 (15.07.2022)/task04.rb:
```diff
@@ -0,0 +1,151 @@
+
+class Hash
+  def with_indifferent_access #: return HashWithIndifferentAccess
+    buf = HashWithIndifferentAccess.new(self)
```
В данном случае, self не имеет смысла передавать в конструктор, так в него мы передаем значение для умолчания. Когда не будет найдено значения по ключу. Но в этом методе достаточно просто указать так: `HashWithIndifferentAccess[self]`

zuwarka replied:
То есть, при записи `HashWithIndifferentAccess.new()` мы просто создали пустой объект класса-наследника, но при `HashWithIndifferentAccess[self]` мы передали в конструктор копию объекта, вызвавшего данный метод? Или же здесь какие-то иные метаморфозы произошли?

Alex6Alex commented on homework04 (15.07.2022)/task04.rb:
```diff
@@ -0,0 +1,151 @@
+class HashWithIndifferentAccess < Hash
+  def [](key) # key == :a => 'a'; key == 'a' => 'a'; real_key = :a
+    buf_key = key.is_a?(Symbol) ? key.to_s : key
+    self.each do |k, v|
```
each метод после выполнения возвращает сам хеш. Так что, если ключ найден не будет, результат поиска по ключу будет сам хеш. Также, не будет работать выбор по ключу, если сам ключ, например, будет числом. В данном методе стоит сделать проверки по типу:

1. Если ключ есть в виде строки - вернуть значение; иначе поискать по ключу в виде символа.
2. Если ключ изначально в виде символа - вернуть значение; иначе поискать по ключу в виде строки.
3. Если ключ не строка и не символ - просто поискать по ключу

Alex6Alex commented on homework04 (15.07.2022)/task04.rb:
```diff
@@ -0,0 +1,151 @@
+  end
+
+  def [](index)
+    vector.to_a[index]
```
Вектор мы задаем в виде массива, так что to_a здесь необязательный

Alex6Alex commented on homework04 (15.07.2022)/task04.rb:
```diff
@@ -0,0 +1,151 @@
+      index += 1
+    end
+    vector
+    # Здесь хотелось бы сделать следующий алгшоритм, но я упорно не понимаю, на что ругается компилятор
```
Ругается, так как переменная из each - это не индекс, а сам элемент

Alex6Alex replied:
Но в этих методах и не очень корректно будет менять сами значения текущего класса. Лучше в виде результата возвращать новый вектор с новым массивом значений. Если менять существующий вектор-класс, то лучше назвать методы со знаком ! на конце

Alex6Alex commented on homework04 (15.07.2022)/task04.rb:
```diff
@@ -0,0 +1,151 @@
+  end
+
+  def to_s
+    vector.to_s
```
Простой to_s не даст того результата, который требуется по условию

zuwarka committed:
Update task04.rb

Исправлены недочеты в классе вектора в методах add!, dot!. to_s; также исправлена перегрузка скобок в классе-наследника Hash

zuwarka committed:
Update task04.rb
